src/lib.rs

```diff
@@ -1128,7 +1131,12 @@ pub unsafe trait HasField<Field, const VARIANT_ID: i128, const FIELD_ID: i128> {
     type Type: ?Sized;
 
     /// Projects from `slf` to the field.
-    fn project(slf: PtrInner<'_, Self>) -> PtrInner<'_, Self::Type>;
+    ///
+    /// # Safety
+    ///
+    /// `slf` must be a non-null pointer, and its referent must be zero bytes in
+    /// size or live in a single allocation.
```

Comment:
This contract isn't strong enough to render projections via `addr_of_mut!` sound, which stipulates:

> The `expr` in `addr_of_mut!(expr)` is evaluated as a place expression, but never loads from the place or requires the place to be dereferenceable. This means that `addr_of_mut!((*ptr).field)` still requires the projection to `field` to be in-bounds, using the same rules as `offset`. However, `addr_of_mut!(*ptr)` is defined behavior even if `ptr` is null, dangling, or misaligned.

For projection, I think we do need something approximating the `PtrInner` guarantee of a valid allocation.